File: platform/pkg/grpc/client/grpc_client.go
// platform/pkg/grpc/client/grpc_client.go
package client

import (
	"context"
	"fmt"
	"sync"

	"google.golang.org/grpc"
	"google.golang.org/grpc/connectivity"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/keepalive"

	"github.com/0xsj/scout/platform/pkg/grpc/interceptors"
)

// grpcClient implements the Client interface using google.golang.org/grpc.
type grpcClient struct {
	config *Config
	conn   *grpc.ClientConn
	mu     sync.RWMutex
}

// New creates a new gRPC client with the given configuration.
func New(config *Config) (Client, error) {
	if config.Logger == nil {
		return nil, fmt.Errorf("logger is required")
	}

	return &grpcClient{
		config: config,
	}, nil
}

// Connect establishes a connection to the gRPC server.
func (c *grpcClient) Connect(ctx context.Context) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.conn != nil {
		return fmt.Errorf("client already connected")
	}

	c.config.Logger.Info("grpc client connecting", "target", c.config.Target)

	// Dial options
	opts := []grpc.DialOption{
		// Keepalive parameters
		grpc.WithKeepaliveParams(keepalive.ClientParameters{
			Time:                c.config.KeepAliveTime,
			Timeout:             c.config.KeepAliveTimeout,
			PermitWithoutStream: true,
		}),

		// Unary interceptors
		grpc.WithChainUnaryInterceptor(
			interceptors.UnaryClientLoggingInterceptor(c.config.Logger),
		),

		// Stream interceptors
		grpc.WithChainStreamInterceptor(
			interceptors.StreamClientLoggingInterceptor(c.config.Logger),
		),
	}

	// TLS or insecure
	if c.config.Insecure {
		opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
	}

	// Create connection with timeout
	dialCtx, cancel := context.WithTimeout(ctx, c.config.ConnectTimeout)
	defer cancel()

	conn, err := grpc.DialContext(dialCtx, c.config.Target, opts...)
	if err != nil {
		return fmt.Errorf("failed to connect to %s: %w", c.config.Target, err)
	}

	c.conn = conn

	c.config.Logger.Info("grpc client connected", "target", c.config.Target)

	return nil
}

// Close closes the connection.
func (c *grpcClient) Close() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.conn == nil {
		return nil
	}

	c.config.Logger.Info("grpc client closing", "target", c.config.Target)

	err := c.conn.Close()
	c.conn = nil

	if err != nil {
		return fmt.Errorf("failed to close connection: %w", err)
	}

	c.config.Logger.Info("grpc client closed")

	return nil
}

// GetConnection returns the underlying grpc.ClientConn.
func (c *grpcClient) GetConnection() *grpc.ClientConn {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.conn
}

// IsConnected returns true if the client is connected.
func (c *grpcClient) IsConnected() bool {
	c.mu.RLock()
	defer c.mu.RUnlock()

	if c.conn == nil {
		return false
	}

	state := c.conn.GetState()
	return state == connectivity.Ready || state == connectivity.Idle
}

File: platform/pylib/scout_common/grpc/client/grpc_client.py
"""gRPC client implementation."""

import time
from typing import Any

import grpc

from scout_common.observability.logger import Logger
from .config import ClientConfig


class GrpcClient:
    """gRPC client implementation using grpc."""
    
    def __init__(self, config: ClientConfig) -> None:
        """
        Initialize gRPC client.
        
        Args:
            config: Client configuration
            
        Raises:
            ValueError: If logger is not provided
        """
        if config.logger is None:
            raise ValueError("logger is required")
        
        self._config = config
        self._logger = config.logger
        self._channel: grpc.Channel | None = None
    
    def connect(self) -> None:
        """Establish a connection to the gRPC server."""
        if self._channel is not None:
            raise RuntimeError("client already connected")
        
        self._logger.info("grpc client connecting", target=self._config.target)
        
        # Channel options
        options = [
            ("grpc.keepalive_time_ms", self._config.keepalive_time_sec * 1000),
            ("grpc.keepalive_timeout_ms", self._config.keepalive_timeout_sec * 1000),
            ("grpc.keepalive_permit_without_calls", 1),
            ("grpc.http2.max_pings_without_data", 0),
        ]
        
        # Create channel
        if self._config.insecure:
            self._channel = grpc.insecure_channel(self._config.target, options=options)
        else:
            # TODO: Add secure channel with credentials
            credentials = grpc.ssl_channel_credentials()
            self._channel = grpc.secure_channel(
                self._config.target,
                credentials,
                options=options,
            )
        
        # Wait for connection to be ready (with timeout)
        try:
            grpc.channel_ready_future(self._channel).result(
                timeout=self._config.connect_timeout_sec
            )
            self._logger.info("grpc client connected", target=self._config.target)
        except grpc.FutureTimeoutError:
            self._channel.close()
            self._channel = None
            raise RuntimeError(
                f"failed to connect to {self._config.target} within {self._config.connect_timeout_sec}s"
            )
    
    def close(self) -> None:
        """Close the connection."""
        if self._channel is None:
            return
        
        self._logger.info("grpc client closing", target=self._config.target)
        
        self._channel.close()
        self._channel = None
        
        self._logger.info("grpc client closed")
    
    def get_channel(self) -> grpc.Channel:
        """
        Get the underlying grpc.Channel.
        
        Returns:
            The grpc.Channel instance
            
        Raises:
            RuntimeError: If client is not connected
        """
        if self._channel is None:
            raise RuntimeError("client not connected")
        
        return self._channel
    
    def is_connected(self) -> bool:
        """Check if the client is connected."""
        if self._channel is None:
            return False
        
        try:
            state = self._channel._channel.check_connectivity_state(False)
            return state == grpc.ChannelConnectivity.READY
        except Exception:
            return False


def new(config: ClientConfig) -> GrpcClient:
    """
    Create a new gRPC client.
    
    Args:
        config: Client configuration
        
    Returns:
        New gRPC client instance
    """
    return GrpcClient(config)